File: src/main/java/Data/JsonUtil.java
package Data;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class JsonUtil {
    private static final JsonUtil instance = new JsonUtil();
    private static final ObjectMapper mapper = new ObjectMapper();
    private static Path dataPath;
    private static Path userImagesPath;

    private JsonUtil() {
        try {
            Path dataPath = Paths.get(System.getProperty("user.home"), ".DnD-Character-Manager");
            if(!dataPath.toFile().exists()){
                Files.createDirectories(dataPath);
                System.out.println("Created data dir: " + dataPath.toAbsolutePath());
            }
            System.out.println(".DnD-Character-Manager exists already");
            JsonUtil.dataPath = dataPath;

            Path userImagesPath = JsonUtil.dataPath.resolve("user_images");
            if(!userImagesPath.toFile().exists()) {
                Files.createDirectories(userImagesPath);
                System.out.println("Created user image dir: " + userImagesPath.toAbsolutePath());
            }
            JsonUtil.userImagesPath = userImagesPath;
        }
        catch (Exception e){
            System.out.println("could not check for .DnD-Character-Manager");
            e.printStackTrace();
        }
    }

    private JsonUtil getInstance(){
        return instance;
    }

    public static Path getDataPath(){
        return dataPath;
    }

    public static Path getUserImagesPathPath(){
        return userImagesPath;
    }

    public static ObjectMapper getMapper() {
        return mapper;
    }
}

File: src/main/kotlin/data/JsonUtil.kt
package data

import main.ItemClasses.Armor
import main.ItemClasses.Consumable
import main.ItemClasses.Item
import main.ItemClasses.Miscellaneous
import main.ItemClasses.Potion
import main.ItemClasses.Weapons.LongRangeWeapon
import main.ItemClasses.Weapons.ShortRangeWeapon
import kotlinx.serialization.json.Json
import kotlinx.serialization.modules.SerializersModule
import kotlinx.serialization.modules.polymorphic
import kotlinx.serialization.modules.subclass
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

object JsonUtil {
    val dataPath: Path
    val userImagesPath: Path

    val json = Json {
        prettyPrint = true
        ignoreUnknownKeys = true
        encodeDefaults = true
        serializersModule = SerializersModule {
            polymorphic(Item::class) {
                subclass(ShortRangeWeapon::class)
                subclass(LongRangeWeapon::class)
                subclass(Consumable::class)
                subclass(Potion::class)
                subclass(Miscellaneous::class)
                subclass(Armor::class)
            }
        }
    }

    init {
        val home = System.getProperty("user.home")
        val dndPath = Paths.get(home, ".DnD-Character-Manager")
        if (!Files.exists(dndPath)) {
            Files.createDirectories(dndPath)
            println("Created data dir: ${dndPath.toAbsolutePath()}")
        } else {
            println(".DnD-Character-Manager exists already")
        }
        dataPath = dndPath

        val imagesPath = dataPath.resolve("user_images")
        if (!Files.exists(imagesPath)) {
            Files.createDirectories(imagesPath)
            println("Created user image dir: ${imagesPath.toAbsolutePath()}")
        }
        userImagesPath = imagesPath
    }
}
